File: java/open_redirect/AccountManagement.java
import jakarta.servlet.http.*;
import java.io.*;

public class AccountManagement extends HttpServlet {

    // VULN 1: Unvalidated `redirect_to` parameter used after password reset confirmation
    public void handlePasswordResetConfirm(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String token      = req.getParameter("token");
        String newPassword = req.getParameter("new_password");
        String redirectTo  = req.getParameter("redirect_to");
        if (validateResetToken(token)) {
            updatePassword(token, newPassword);
            resp.sendRedirect(redirectTo);
        } else {
            resp.sendRedirect("/reset-password?error=invalid_token");
        }
    }

    // VULN 2: Referer header used directly as logout redirect destination
    public void handleLogout(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String referer = req.getHeader("Referer");
        invalidateSession(req.getSession());
        if (referer != null && !referer.isEmpty()) {
            resp.sendRedirect(referer);
        } else {
            resp.sendRedirect("/");
        }
    }

    // VULN 3: Unvalidated `callback_url` used after social account linking
    public void handleSocialAccountLink(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String provider    = req.getParameter("provider");
        String oauthToken  = req.getParameter("oauth_token");
        String callbackUrl = req.getParameter("callback_url");
        linkSocialAccount(provider, oauthToken);
        resp.sendRedirect(callbackUrl);
    }

    private boolean validateResetToken(String token) { return true; }
    private void updatePassword(String token, String password) {}
    private void invalidateSession(HttpSession session) { session.invalidate(); }
    private void linkSocialAccount(String provider, String token) {}
}

File: kotlin/open_redirect/AccountManagement.kt
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class AccountManagement {

    // VULN 1: response.sendRedirect with unvalidated redirectTo in password reset flow
    @GetMapping("/account/reset-password/confirm")
    fun confirmPasswordReset(
        @RequestParam token: String,
        @RequestParam newPassword: String,
        @RequestParam(defaultValue = "/account/login") redirectTo: String,
        response: HttpServletResponse
    ) {
        response.sendRedirect(redirectTo)
    }

    // VULN 2: Referer header used directly as redirect target after logout
    @PostMapping("/auth/logout")
    fun logout(
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        request.getSession(false)?.invalidate()
        val referer = request.getHeader("Referer") ?: "/shop"
        response.sendRedirect(referer)
    }

    // VULN 3: callback_url param used directly as redirect for social account linking
    @GetMapping("/account/social/link/callback")
    fun socialLinkCallback(
        @RequestParam provider: String,
        @RequestParam code: String,
        @RequestParam(defaultValue = "/account/settings") callback_url: String,
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        val session = request.getSession(false)
        session?.setAttribute("${provider}_linked", true)
        response.sendRedirect(callback_url)
    }
}
